### PR TITLE
[FINERACT-1965] Change date resolvement from SMART to STRICT

### PR DESCRIPTION
**Pull request by taskain7**
## Description

Changing date resolvement from SMART to STRICT
DateFormat value object has been indtroduced.
DateFormat is used for query parameters at soma API-s.
Extended usage of DateFormat will be coming in further PRs.

### CODE REVIEW

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java [hunk 1]
```diff
@@ -518,17 +519,19 @@ public static LocalDate convertFrom(final String dateAsString, final String para
         return convertDateTimeFrom(dateAsString, parameterName, dateFormat, clientApplicationLocale).toLocalDate();
     }
 
-    public static LocalDateTime convertDateTimeFrom(final String dateTimeAsString, final String parameterName, final String dateTimeFormat,
+    public static LocalDateTime convertDateTimeFrom(final String dateTimeAsString, final String parameterName, String dateTimeFormat,
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:536]: `extractLocalDateTimeNamed` was missed. I reckon that method should be also enhanced.

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java [hunk 2]
```diff
@@ -518,17 +519,19 @@ public static LocalDate convertFrom(final String dateAsString, final String para
...
             final Locale clientApplicationLocale) {
 
         validateDateFormatAndLocale(parameterName, dateTimeFormat, clientApplicationLocale);
         LocalDateTime eventLocalDateTime = null;
+        dateTimeFormat = dateTimeFormat.replace("y", "u");
         if (StringUtils.isNotBlank(dateTimeAsString)) {
             try {
                 DateTimeFormatter formatter = new DateTimeFormatterBuilder().parseCaseInsensitive().parseLenient()
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:544]: Should we remove the `parseLenient()` here?

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java [hunk 3]
```diff
@@ -518,17 +519,19 @@ public static LocalDate convertFrom(final String dateAsString, final String para
         return convertDateTimeFrom(dateAsString, parameterName, dateFormat, clientApplicationLocale).toLocalDate();
     }
 
-    public static LocalDateTime convertDateTimeFrom(final String dateTimeAsString, final String parameterName, final String dateTimeFormat,
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:521]: `extractMonthDayNamed` was missed. I reckon that method should be also enhanced.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java (hunk 3 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:521]: fixed

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java (hunk 1 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:536]: fixed

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java (hunk 2 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:544]: I think we shouldn't, parsing could remain lenient, it's still useful in cases like "9 September 2022" will be parsed to "09 September 2022"

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        this.requestSpec = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        this.requestSpec.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        this.responseSpec = new ResponseSpecBuilder().expectStatusCode(200).build();
+        this.errorResponseSpec = new ResponseSpecBuilder().expectStatusCode(400).build();
+        this.clientHelper = new ClientHelper(this.requestSpec, this.responseSpec);
+        this.loanTransactionHelper = new LoanTransactionHelper(requestSpec, responseSpec);
+    }
+
+    @Test
+    public void testShouldFailIfDateIsInvalid() {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java:91]: Please add 2 more test cases which covers `extractLocalTimeNamed` and `extractMonthDayNamed` method changes as well.

Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java:91]: done

File: fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java [hunk 5]
```diff
@@ -26,7 +26,7 @@ private InteropUtil() {
 
     }
 
-    public static final String ISO8601_DATE_TIME_FORMAT = "yyyy-MM-ddTHH:mm:ss.SSS[-HH:MM]";
+    public static final String ISO8601_DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss.SSS";
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: Why did you change this? It looks breaking change!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    @Test
+    public void testShouldFailWithInvalidMonthDay() {
+        final Account assetAccount = this.accountHelper.createAssetAccount();
+        final Account incomeAccount = this.accountHelper.createIncomeAccount();
+        final Account expenseAccount = this.accountHelper.createExpenseAccount();
+        final Account liabilityAccount = this.accountHelper.createLiabilityAccount();
+
+        DateFormat dateFormat = new SimpleDateFormat("dd MMMM yyyy", Locale.US);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: Please use DateTimeFormatter!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        new SimpleDateFormat("dd MMM", Locale.US);
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: this seems useless :)

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        DateFormat currentDateFormat = new SimpleDateFormat("dd");
+
+        Calendar todaysDate = Calendar.getInstance();
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: Please use LocalDate instead of Calendar...

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        todaysDate.add(Calendar.MONTH, -3);
+        final String VALID_FROM = dateFormat.format(todaysDate.getTime());
+        todaysDate.add(Calendar.YEAR, 10);
+        final String VALID_TO = dateFormat.format(todaysDate.getTime());
+
+        todaysDate = Calendar.getInstance();
+        todaysDate.add(Calendar.MONTH, -1);
+        final String SUBMITTED_ON_DATE = dateFormat.format(todaysDate.getTime());
+
+        Integer currentDate = Integer.valueOf(currentDateFormat.format(todaysDate.getTime()));
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: Unused calculations [161-164]

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        Integer daysInMonth = todaysDate.getActualMaximum(Calendar.DATE);
+        Integer numberOfDaysLeft = daysInMonth - currentDate + 1;
+        todaysDate.add(Calendar.DATE, numberOfDaysLeft);
+
+        PostClientsRequest postClientsRequest = ClientHelper.defaultClientCreationRequest();
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java:158]: Please remove all the parts that are not relevant for the test case!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java (hunk 5 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: but it cannot be parsed with SMART resolvement

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: you mean STRICT?

Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: yes, sorry

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java (hunk 10 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java:158]: I need a client ID for the test request

Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java:158]: that was more of a general request, not particularly for this line ;)

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java (hunk 9 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: removed

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java (hunk 8 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: fixed

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java (hunk 7 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: removed

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java (hunk 6 above)
Comment (taskain7) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DateValidationTest.java]: fixed

File: fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java (hunk 5 above)
Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: It's under discussion wether or not we can change it.

Comment (taskain7) [reply fineract-provider/src/main/java/org/apache/fineract/interoperation/util/InteropUtil.java]: the pattern was incorrect, has been fixed

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DateFormat.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+import lombok.Getter;
+import org.apache.fineract.infrastructure.core.exception.PlatformApiDataValidationException;
+
+public class DateFormat {
+
+    @Getter
+    private final String dateFormat;
+
+    public DateFormat(String rawDateFormat) {
+        String compatibleDateFormat = rawDateFormat.replace("yyyy", "uuuu");
```
Comment (galovics) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DateFormat.java]: Let's add a nullcheck before here as well.

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java [hunk 12]
```diff
@@ -459,28 +462,31 @@ public LocalTime extractLocalTimeNamed(final String parameterName, final JsonEle
...
-        String timeValueAsString = null;
+        String timeValueAsString;
         if (element.isJsonObject()) {
             final JsonObject object = element.getAsJsonObject();
             if (object.has(parameterName) && object.get(parameterName).isJsonPrimitive()) {
                 parametersPassedInCommand.add(parameterName);
 
                 try {
-                    DateTimeFormatter timeFormtter = DateTimeFormatter.ofPattern(timeFormat);
+                    timeFormat = timeFormat.replace("y", "u");
```
Comment (galovics) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:476]: I don't think we need this anymore, do we?

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java [hunk 13]
```diff
@@ -518,17 +524,25 @@ public static LocalDate convertFrom(final String dateAsString, final String para
...
+    }
+
+    private static LocalDateTime convertDateTimeFrom(final String dateTimeAsString, final String parameterName, String dateTimeFormat,
             final Locale clientApplicationLocale) {
 
         validateDateFormatAndLocale(parameterName, dateTimeFormat, clientApplicationLocale);
         LocalDateTime eventLocalDateTime = null;
         if (StringUtils.isNotBlank(dateTimeAsString)) {
             try {
+                dateTimeFormat = dateTimeFormat.replace("y", "u");
```
Comment (galovics) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:543]: Same as above.

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DateFormat.java (hunk 11 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/data/DateFormat.java]: added

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java (hunk 12 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:476]: it's still necessary since I haven't changed the type everywhere. The rest will be coming in a furhter PR.

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java (hunk 13 above)
Comment (taskain7) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:543]: it's still necessary since I haven't changed the type everywhere. The rest will be coming in a furhter PR.

**Review by galovics — APPROVED** [review]
